### PR TITLE
New healthcheck

### PR DESCRIPTION
**Pull request by bludnic**
- [x] Added unified Healthcheck for all nodes (ADM, BTC, ETH, DOGE, DASH, LSK)
- [x] Show additional info for "Unsupported status": `Http is not allowed`, `Outdated API version`, `Not reachable`

image

### CODE REVIEW

File: src/lib/nodes/abstract.node.ts [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+   */
+  port: string
+  /**
+   * Node hostname like bid.adamant.im or 23.226.231.225
+   */
+  hostname: string
+  /**
+   * WebSocket protocol
+   */
+  wsProtocol: 'ws:' | 'wss:' = 'wss:'
```
[the remark this answers is not in the document]
Comment (bludnic) [reply src/lib/nodes/abstract.node.ts:53]: This code `new URL('wss://hello.com').protocol` will return `wss:`. You can see a similar result with `https:` in the constructor.

File: src/components/NodesTable/NodesTableItem.vue [hunk 2]
```diff
@@ -11,35 +11,40 @@
...
-      />
+    <td :class="classes.td" class="pl-0 pr-2" :colspan="unsupported ? 2 : 1">
+      <div>
+        <span>
+          {{ nodeStatus }}
+        </span>
+        <v-icon
+          :class="{
+            [classes.statusIcon]: true,
+            [classes.statusIconGreen]: nodeStatusColor === 'green',
```
[the remark this answers is not in the document]
Comment (bludnic) [reply src/components/NodesTable/NodesTableItem.vue:34]: We don't use computed string for class names

File: src/lib/nodes/abstract.client.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
+import { filterSyncedNodes } from './utils/filterSyncedNodes.ts'
+import { Node } from './abstract.node'
```
[the remark this answers is not in the document]
Comment (bludnic) [reply src/lib/nodes/abstract.client.ts:2]: removed `.ts` extenstion

File: src/lib/nodes/adm/AdmNode.ts [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+  }
+
+  /**
+   * Performs an API request.
+   *
+   * The `payload` of the `cfg` can be either an object or a function that
+   * accepts `ApiNode` as a first argument and returns an object.
+   */
+  request<P extends Payload = Payload>(cfg: RequestConfig<P>) {
+    const { url, method = 'get', payload } = cfg
```
[the remark this answers is not in the document]
Comment (bludnic) [reply src/lib/nodes/adm/AdmNode.ts:54]: I think it's ok to destruct in function body

File: src/lib/nodes/btc/BtcNode.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+ * to the node and verify is status (online/offline, version, ping, etc.)
+ */
+export class BtcNode extends Node {
+  client: AxiosInstance
+
+  constructor(url: string) {
+    super(url)
+
+    this.client = axios.create({ baseURL: url })
+    this.client.interceptors.response.use(null, (error) => {
```
[the remark this answers is not in the document]
Comment (bludnic) [reply src/lib/nodes/btc/BtcNode.ts]: added `createBtcLikeClient` helper

File: src/store/modules/nodes/nodes-plugin.js [hunk 6]
```diff
@@ -1,22 +1,45 @@
...
+import { adm } from '@/lib/nodes/adm'
+import { eth } from '@/lib/nodes/eth'
+import { btc } from '@/lib/nodes/btc'
+import { doge } from '@/lib/nodes/doge'
+import { dash } from '@/lib/nodes/dash'
+import { lsk } from '@/lib/nodes/lsk'
+
+export default (store) => {
+  // initial nodes state
+  adm.getNodes().forEach((status) => store.commit('nodes/status', { status, nodeType: 'adm' }))
```
[the remark this answers is not in the document]
Comment (bludnic) [reply src/store/modules/nodes/nodes-plugin.js:10]: To iterate all nodes in one place we would need something like `NodesManager` class. If we had 100 nodes it make sense. The current approach is ok. There is no need in another abstraction.

[the remark this answers is not in the document]
Comment (bludnic) [reply src/store/modules/nodes/nodes-plugin.js]: I use both named, and default export. See Vuex crypto modules

File: src/store/modules/nodes/nodes-actions.js [hunk 7]
```diff
@@ -1,21 +1,31 @@
...
+
+export default {
+  restore({ state }) {
+    const nodes = Object.values(state.adm)
+
+    nodes.forEach((node) => adm.toggleNode(node.url, node.active))
+  },
+
+  updateStatus() {
+    adm.checkHealth()
```
[the remark this answers is not in the document]
Comment (bludnic) [reply src/store/modules/nodes/nodes-actions.js]: fixed

**Comment by bludnic** [comment]
@adamant-al check the UI for coin nodes. Does it look ok?

**Comment by adamant-al** [comment]
1. Add infoService health check
2. Check `alt_ip` if web domain is not accessible
3. Rename `Nodes list` to `ADM nodes`
4. Move coin nodes to the `Coin nodes` and place it after `ADM nodes` in the Settings
5. Add the `Service nodes` for infoService and place it after `Coin nodes` in the Settings
6. Change the padding

Screenshot 2023-11-05 в 17 19 00

7. Remove `Socket` column in the `Coin nodes` screen
8. Move coin name suffix to a separate, the first column, same text format
9. Suffixes: eth-node, eth-indexer, btc-node, btc-indexer, doge-node, doge-indexer, dash-node, lsk-node, lsk-indexer, rates-info. Consider to include them in the `adamant-wallets` repo.
10. Sort by suffix ascending
11. Show coin node version where available
12. Under the ping value, show node's height
13. For infoService, it's `last_update` timestamp — to be added to https://info.adamant.im/get. Show as '13 minutes ago'. Threshold is 30 min (set a constant).
14. Add health check for indexers
15. Later we'll allow disabling coin nodes (not to expose user IP)

**Comment by bludnic** [comment]
Created a task https://trello.com/c/Uhj8LOdD/456-healthcheck-improvements

**Comment by bludnic** [comment]
@adamant-al fixed "6. Change the padding"

image

For everything else I created a task for future release: https://trello.com/c/Uhj8LOdD/456-healthcheck-improvements

**Comment by bludnic** [comment]
@adamant-al need your approval

**Comment by bludnic** [comment]
I will fix in the current PR: 3, 4, 6, 7, 8, 9, 10, 11, 12.
Other changes related to indexers/services healthcheck will be done in https://trello.com/c/Uhj8LOdD/456-healthcheck-improvements

**Comment by bludnic** [comment]
image
image

**Comment by bludnic** [comment]
- Made `NodesTable` reusable. Decomposed into smaller components
- Split `NodesTable` into `AdmNodeTable` and `CoinNodesTable`
- Added label (suffix) to the Node instance

**Comment by bludnic** [comment]
[Feat: Health check for coin nodes and infoservices](https://trello.com/c/MQVhzbjO/232-feat-health-check-for-coin-nodes-and-infoservices)

**Comment by bludnic** [comment]
Seems like node's height takes up too much space. Proposed solution:

image

Icon: https://pictogrammers.com/library/mdi/icon/cube-outline/

@adamant-al

**Comment by adamant-al** [comment]
Added the `Format the Nodes screen (screenshots)` list item in the Feat: Health check improvements https://trello.com/c/Uhj8LOdD
You can do a part here.
No tooltips.

**Comment by bludnic** [comment]
- Added cube-icon before node height
- Fixed conflicts

File: src/components/nodes/adm/AdmNodesTable.vue [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+      classes
+    }
+  }
+})
+</script>
+
+<style lang="scss">
+@import 'vuetify/settings';
+
+.adm-nodes-table {
```
Comment (RealGoodProgrammer) [inline src/components/nodes/adm/AdmNodesTable.vue:49]: Unnecessary

**Review by RealGoodProgrammer — APPROVED** [review]

**Pull request merged**